src/creature.cpp

```diff
@@ -1587,20 +1588,22 @@ body_part Creature::select_body_part(Creature *source, int hit_roll)
     std::set<std::pair<body_part, double>, weight_compare> adjusted_weights;
     for(iter = hit_weights.begin(); iter != hit_weights.end(); ++iter) {
         totalWeight += iter->second;
-        adjusted_weights.insert(*iter);
+        adjusted_weights.insert( *iter );
     }
 
-    double roll = rng_float(1, totalWeight);
-    body_part selected_part = bp_torso;
-
-    std::set<std::pair<body_part, double>, weight_compare>::iterator adj_iter;
-    for(adj_iter = adjusted_weights.begin(); adj_iter != adjusted_weights.end(); ++adj_iter) {
-        roll -= adj_iter->second;
-        if(roll <= 0) {
-            selected_part = adj_iter->first;
+    double roll = rng_float(0, totalWeight);
+    body_part selected_part;
+    do {
+        auto adj_iter = random_entry_removed( adjusted_weights );
```

Comment:
To put it another way, since we're choosing a weight up front, iterating through the list in a random order doesn't make things any better.